app/javascript/packages/transliterable-field-group/index.ts

```diff
@@ -0,0 +1,375 @@
+
+/**
+ * Representation of the response from a transliteration API call
+ */
+interface ValidationResponse {
+  success: boolean;
+  data: Record<string, string>;
+}
+
+/**
+ * Internal cache structure for controlling when API call is made
+ * and reusing result
+ */
+interface ValidationResultCache {
+  preValidationValues: Record<string, string>;
+  postValidationValues: Record<string, string>;
+  result: Record<string, string>;
+}
+
+/**
+ * This adds an asynchronous transliteration check for certain form fields
+ * that gets run between form validation and submission.
+ */
+class TransliterableFieldGroupElement extends HTMLElement {
+  private static readonly DEFAULT_INPUT_TIMEOUT_MS = 1500;
+
+  /**
+   * Cache to moderate API calls and allow reuse of the previous result
+   */
+  private validationResultCache: ValidationResultCache = {
+    preValidationValues: {},
+    postValidationValues: {},
+    result: {},
+  };
+
+  private lastRevalidationRequest: Promise<void>;
+
+  private lastFormSubmit: Promise<any> | null = null;
+
+  /**
+   * Form associated with this element
+   */
+  get form(): HTMLFormElement | null {
+    let element: HTMLElement | null = this;
+    do {
+      element = element.parentElement;
+    } while (element && !(element instanceof HTMLFormElement));
+    return element;
+  }
+
+  /**
+   * URL to fetch from for validating that fields are transliterable
+   */
+  get validationUrl(): string | null {
+    return this.getAttribute('validation-url');
+  }
+
+  /**
+   * Time to wait for input to stop before validating for transliterability
+   */
+  get inputTimeoutMilliseconds(): number {
+    const rawAttribute = this.getAttribute('input-timeout-ms');
+    if (rawAttribute === null) {
+      return TransliterableFieldGroupElement.DEFAULT_INPUT_TIMEOUT_MS;
+    }
+
+    const attrNumber = Number(rawAttribute);
+    if (Number.isNaN(attrNumber) || attrNumber < 0) {
+      return TransliterableFieldGroupElement.DEFAULT_INPUT_TIMEOUT_MS;
+    }
+    return attrNumber;
+  }
+
+  /**
+   * Element that displays general form errors to the user
+   */
+  get formErrorElement(): Element | null {
+    const selector = this.getAttribute('form-error-selector');
+    if (selector) {
+      return this.form?.querySelector(selector) || null;
+    }
+
+    return null;
+  }
+
+  /**
+   * Mapping of regular field names to transliteration API field names
+   */
+  get fieldMapping(): Record<string, string> {
+    const rawFields = this.getAttribute('field-mapping');
+    if (rawFields) {
+      try {
+        return JSON.parse(rawFields);
+      } catch (e) {
+        trackError(e);
+      }
+    }
+    return {};
+  }
+
+  /**
+   * Input elements mapped by their transliteration API field names
+   */
+  private get inputs(): Record<string, HTMLInputElement> {
+    // Extract transliterable fields
+    const { fieldMapping, form } = this;
+    return Object.entries(fieldMapping).reduce((agg, [key, value]) => {
```

Comment:
what is agg?

Reply:
`agg` here is the typical "aggregate" value in a `reduce` call. In this case we're doing the following transformation.

1. Getting the field mapping value specified on the current element's attribute: `<lg-transliterable-field-group field-mapping="(JSON encoded object)">...`
2. Convert the object to an array of `[key, value]` entries
3. Use the `key` to fetch the associated `<input>` elements from the form
4. Set each input as a value on `agg` accessible by a key of `value`

That results here in an object containing the transliterable `HTMLInputElement`s accessible using the "short" name for the associated values in the transliteration API.